pkg/auth/requests/service_account.go

```diff
@@ -0,0 +1,161 @@
+package requests
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+
+	jwtv4 "github.com/golang-jwt/jwt/v4"
+	"github.com/gorilla/mux"
+	authcontext "github.com/rancher/rancher/pkg/auth/context"
+	"github.com/rancher/rancher/pkg/auth/tokens"
+	controllers "github.com/rancher/rancher/pkg/generated/controllers/management.cattle.io/v3"
+	corev1 "github.com/rancher/rancher/pkg/generated/norman/core/v1"
+	mgmtv3 "github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/types/config"
+	"github.com/rancher/steve/pkg/auth"
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/api/authentication/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	authv1 "k8s.io/client-go/kubernetes/typed/authentication/v1"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	serviceAccountSubjectPrefix = "system:serviceaccount:"
+	settingsObjectName          = "clusterproxyconfig"
+)
+
+type (
+	restConfigGetter  func(cluster *mgmtv3.Cluster, context *config.ScaledContext, secretLister corev1.SecretLister) (*rest.Config, error)
+	authClientCreator func(config *rest.Config) (authv1.AuthenticationV1Interface, error)
+)
+
+// ServiceAccountAuth is an authenticator that authenticates requests using the downstream service account's JWT.
+type ServiceAccountAuth struct {
+	scaledContext             *config.ScaledContext
+	clusterLister             mgmtv3.ClusterLister
+	secretLister              corev1.SecretLister
+	restConfigGetter          restConfigGetter
+	authClientCreator         authClientCreator
+	clusterProxyConfigsGetter controllers.ClusterProxyConfigCache
+}
+
+// NewServiceAccountAuth creates a new instance of ServiceAccountAuth.
+func NewServiceAccountAuth(
+	scaledContext *config.ScaledContext,
+	restConfigGetter restConfigGetter,
+) auth.Authenticator {
+	return &ServiceAccountAuth{
+		scaledContext:    scaledContext,
+		clusterLister:    scaledContext.Management.Clusters("").Controller().Lister(),
+		secretLister:     scaledContext.Core.Secrets("").Controller().Lister(),
+		restConfigGetter: restConfigGetter,
+		authClientCreator: func(config *rest.Config) (authv1.AuthenticationV1Interface, error) {
+			return authv1.NewForConfig(config)
+		},
+		clusterProxyConfigsGetter: scaledContext.Wrangler.Mgmt.ClusterProxyConfig().Cache(),
+	}
+}
+
+// Authenticate the request using the downstream service account's JWT.
+func (t *ServiceAccountAuth) Authenticate(req *http.Request) (user.Info, bool, error) {
+	// Basic checks to see if we even need to proceed
+	info, hasAuth := request.UserFrom(req.Context())
+	if info.GetName() != "system:cattle:error" {
```

Comment:
This is not the first authenticator. Initially, it attempts authentication with the [tokenAuthenticator](https://github.com/rancher/rancher/blob/release/v2.9/pkg/auth/requests/authenticate.go#L99). If it fails, then it tries the serviceAccount authenticator. Therefore, this condition is necessary.

I've run it with the debugger, and it always tries to authenticate with the rancher token first (even if the sa jwt is provided). This is what I saw:

For rancher token requests:
- [tokenAuthenticator](https://github.com/rancher/rancher/blob/release/v2.9/pkg/auth/requests/authenticate.go#L99) authenticates the user
- serviceAccount authenticator is called but returns [here](https://github.com/pmatseykanets/rancher/blob/service-account-auth/pkg/auth/requests/service_account.go#L64)
- impersonation Authenticate is called

For service account tokens:
- [tokenAuthenticator](https://github.com/rancher/rancher/blob/release/v2.9/pkg/auth/requests/authenticate.go#L99) is called, but it fails to authenticate the user
- new serviceAccount authenticator authenticates the user
- impersonation Authenticate is called, but it does nothing

Reply:
This means the new `serviceAccount` authenticator will be called for all requests to `/k8s/clusters/{clusterID}` , even if it is disabled (e.g. there is no clusterproxyconfig). It exits early, but we might need to consider its impact on performance if we anticipate a high volume of calls.

I might be missing something here, but have you consider creating a new endpoint for sa token authentication? Something like `/k8s/sa/clusters/{clusterID}`. This endpoint would have just the sa authenticator. I believe this would make the code easier to read, and we would not have this extra call in all requests.

Reply:
I hear your concern here.  There was one iteration early-on where we experimented with a separate URL, but we went this direction instead.
I think the expectation here was to use the same endpoint since that is the "normal" way that users currently use to proxy their requests to downstream clusters.  Our goal here is to just make SA tokens work in addition to Rancher tokens and to make it fairly seamless for users without needing to consider that they might need a different URL for different auth modes.